### ISSUE TITLE
Performance bottleneck: Only 2.25 TFLOPS/s due to inefficient scatter/index operations (~26% of CUDA time)

### ISSUE THREAD

**Issue by AmitMY** [sign/WeLT#53]
## Problem Summary

Profiling WeLT training reveals **significant performance bottlenecks** causing only **2.25 TFLOPS/s** GPU utilization - approximately **2% of theoretical peak** for modern GPUs (100+ TFLOPS for BF16).

## Profiling Results

**Setup:** batch_size=32, gemma-3-270m latent transformer, bert-tiny encoder, 30 steps with PyTorch profiler

```
FLOPS Profile @ Step 30
======================================================================
  Compute: 330335.58 GFLOPS | 2.25 TFLOPS/s
  Time:    CUDA 146746.74ms | CPU 119293.83ms

  Top Operations by CUDA Time:
     1. aten::mul                                14762.535ms ( 10.1%)   449.1 GFLOPS
     2. aten::_index_put_impl_                   12272.235ms (  8.4%)        -      
     3. aten::_efficient_attention_backward       9025.130ms (  6.2%)        -      
     4. void (anonymous namespace)::indexing_bac  8151.246ms (  5.6%)        -      
     5. fmha_cutlassB_bf16_aligned_128x64_k65536  7539.797ms (  5.1%)        -      
     6. aten::copy_                               6976.770ms (  4.8%)        -      
     7. aten::mm                                  6660.381ms (  4.5%) 324764.0 GFLOPS
     8. aten::fill_                               5981.399ms (  4.1%)        -      
     9. void at::native::elementwise_kernel       4963.188ms (  3.4%)        -      
    10. void (anonymous namespace)::indexing_bac  4111.889ms (  2.8%)        -      
```

**Key Observations:**
- `aten::mm` (actual matrix multiplication) is only **4.5% of CUDA time** but achieves **324 TFLOPS** - proving the GPU CAN perform well
- **~26% of time** spent on scatter/index operations (`_index_put_impl_`, `indexing_backward`, `copy_`, `fill_`)
- `aten::mul` at **10.1%** achieves only **449 GFLOPS** - suggesting fragmented/non-contiguous tensor operations
- CPU time (119s) is comparable to CUDA time (147s) - potential CPU bottleneck in data preparation

## Batch Size Scaling Test

| Metric | Batch=32 | Batch=8 | Ratio |
|--------|----------|---------|-------|
| TFLOPS/s | 2.25 | 2.31 | 1.03x |
| CUDA time | 146.7s | 35.7s | 4.1x |
| CPU time | 119.3s | 28.2s | 4.2x |

**Same TFLOPS/s regardless of batch size** confirms the bottleneck is architectural, not batch-size related.

### 1. `collator.py:stack_pad_tensors_fast()` - Likely source of `_index_put_impl_`

```python
def stack_pad_tensors_fast(tensors: list[torch.Tensor], pad_value=0, dtype=None, device=None):
    nt = torch.nested.nested_tensor(tensors, dtype=dtype, device=device)
    return nt.to_padded_tensor(pad_value)  # <-- Scatter ops here
```

`to_padded_tensor()` from NestedTensor is known to cause scatter operations as it fills a dense tensor from ragged data.

**Suggestion:** Pre-allocate padded tensor and use vectorized indexing, or batch tensors by similar sizes to reduce padding overhead.

### 2. `attention.py:get_attention_mask_for_packed_sequence()` - Loop-based scatter

```python
def get_attention_mask_for_packed_sequence(seq_lengths: list[int], ...):
    mask = torch.zeros((1, total_length, total_length), dtype=torch.bool)
    for length in seq_lengths:
        mask[0, start_position:end_position, start_position:end_position] = tril[:length, :length]
        # ^-- Scatter operation in a loop
```

**Suggestion:** Compute block-diagonal mask in a single vectorized operation using `torch.block_diag()` or pre-compute masks for common sequence length patterns.

### 3. `model.py:parallel_causal_decode()` - Heavy tensor reshaping

```python
def parallel_causal_decode(self, latent_vectors, target_ids, target_mask):
    B, L, hidden_dim = latent_vectors.shape
    T = target_ids.shape[-1]
    
    # Multiple reshape operations every forward pass:
    target_ids_flat = target_ids.view(B * L, T)  
    target_embeds = embed_layer(target_ids_flat)
    latent_vectors_flat = latent_vectors.view(B * L, hidden_dim).unsqueeze(1)
    combined_embeds = torch.cat([latent_vectors_flat, target_embeds], dim=1)
    # ... more reshapes on output
```

**Suggestion:** Consider fusing operations or restructuring to avoid repeated reshape/view/cat operations. The (B, L, T) → (B*L, T) pattern could potentially be avoided by using 3D batched operations.

### 4. `model.py:encode_texts()` - Non-contiguous indexing

```python
def encode_texts(self, input_ids, attention_mask):
    # ...
    text_embeds = text_embeds[:, 0]  # Index operation
    return text_embeds.view(B, L, -1)  # View after indexing
```

**Suggestion:** Use `.contiguous()` after indexing if the tensor is used in subsequent operations, or consider alternative approaches like pooling.

## Environment

- PyTorch 2.10.0a0 (NVIDIA container 25.12)
- CUDA 13.1 (forward compat mode)
- Profile settings: `flops_warmup_steps=10`, `flops_active_steps=10`

## Suggested Improvements

1. **Replace NestedTensor padding** with pre-allocated dense tensors + masking
2. **Vectorize attention mask creation** instead of loop-based slice assignment
3. **Reduce tensor reshaping** in `parallel_causal_decode()` - consider batched 3D operations
4. **Add `torch.compile()`** for the forward pass to allow kernel fusion
5. **Profile dataloader** separately - CPU time is high, suggesting potential data prep bottleneck

## Reproduction

```yaml
# welt.yaml
profile_flops: true
flops_profile_steps: 10
flops_warmup_steps: 10
flops_active_steps: 10
max_steps: 30
per_device_train_batch_size: 32
```

```bash
docker run --rm --gpus all \
  --ipc=host --ulimit memlock=-1 --ulimit stack=67108864 \
  -v /shared/.cache:/root/.cache \
  -v "$(pwd)":/video-gpt \
  -v "$(pwd)/output":/output \
  -e CONFIG="/video-gpt/welt.yaml" \
  welt
```
